backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/document/model/dsl/impl/ElasticsearchStringImplicitFieldContributor.java

```diff
@@ -0,0 +1,83 @@
+import org.hibernate.search.engine.backend.document.model.dsl.spi.ImplicitFieldContributor;
+import org.hibernate.search.engine.search.aggregation.spi.AggregationTypeKeys;
+import org.hibernate.search.engine.search.predicate.spi.PredicateTypeKeys;
+import org.hibernate.search.engine.search.projection.spi.ProjectionTypeKeys;
+import org.hibernate.search.engine.search.sort.spi.SortTypeKeys;
+
+public class ElasticsearchStringImplicitFieldContributor implements ImplicitFieldContributor {
+
+	private final String name;
+
+	public ElasticsearchStringImplicitFieldContributor(String name) {
+		this.name = name;
+	}
+
+	public void contribute(ImplicitFieldCollector collector) {
+		collector.addImplicitField( name, new ElasticsearchIndexValueFieldType<>( createTypeBuilder() ) );
+	}
+
+	private ElasticsearchIndexValueFieldType.Builder<String> createTypeBuilder() {
+		PropertyMapping mappedTypePropertyMapping = new PropertyMapping();
+		mappedTypePropertyMapping.setType( DataTypes.KEYWORD );
+		mappedTypePropertyMapping.setIndex( false );
+		mappedTypePropertyMapping.setDocValues( true );
+
+		ElasticsearchIndexValueFieldType.Builder<String> typeBuilder = new ElasticsearchIndexValueFieldType.Builder<>(
+				String.class, mappedTypePropertyMapping );
+		typeBuilder.codec( ElasticsearchStringFieldCodec.INSTANCE );
+
+		typeBuilder.searchable( true );
+		typeBuilder.queryElementFactory(
+				PredicateTypeKeys.MATCH, new ElasticsearchTextMatchPredicate.Factory( typeBuilder.codec() ) );
+		typeBuilder.queryElementFactory(
+				PredicateTypeKeys.RANGE, new ElasticsearchRangePredicate.Factory<>( typeBuilder.codec() ) );
+		typeBuilder.queryElementFactory( PredicateTypeKeys.EXISTS, new ElasticsearchExistsPredicate.Factory<>() );
+		typeBuilder.queryElementFactory( PredicateTypeKeys.PHRASE, new ElasticsearchTextPhrasePredicate.Factory() );
+		typeBuilder.queryElementFactory( PredicateTypeKeys.WILDCARD, new ElasticsearchTextWildcardPredicate.Factory() );
+		typeBuilder.queryElementFactory( PredicateTypeKeys.REGEXP, new ElasticsearchTextRegexpPredicate.Factory() );
+		typeBuilder.queryElementFactory(
+				PredicateTypeKeys.TERMS, new ElasticsearchTermsPredicate.Factory<>( typeBuilder.codec() ) );
+		typeBuilder.queryElementFactory(
+				ElasticsearchPredicateTypeKeys.SIMPLE_QUERY_STRING,
+				new ElasticsearchSimpleQueryStringPredicateBuilderFieldState.Factory()
+		);
+		typeBuilder.sortable( true );
+		typeBuilder.queryElementFactory(
+				SortTypeKeys.FIELD, new ElasticsearchStandardFieldSort.TextFieldFactory( typeBuilder.codec() ) );
+
+		typeBuilder.projectable( true );
+		typeBuilder.queryElementFactory(
+				ProjectionTypeKeys.FIELD, new ElasticsearchFieldProjection.Factory<>( typeBuilder.codec() ) );
+
+		typeBuilder.aggregable( true );
+		typeBuilder.queryElementFactory(
+				AggregationTypeKeys.TERMS, new ElasticsearchTermsAggregation.Factory<>( typeBuilder.codec() ) );
+
+		return typeBuilder;
+	}
```

Comment:
This looks like code duplication, as you're manually definining which predicates/projections/etc are available, while that's generally inferred from the type + flags (searchable/projectable/etc.). As types evolve, we're likely to forget to update this code.

Can'y you pass the type factory to implicit field contributors instead, so that they use the API to build a type?

Reply:
Ok, should be a bit cleaner now 😃